File: BreadLang/Parser/Nodes/RegisterNode.cs
using BreadLang.Compiling;
using BreadLang.Tokens;

namespace BreadLang.Parser.Nodes;

public class RegisterNode : Node
{
    public string Register { get; private set; }

    public override void Populate(Parser parser)
    {
        var token = parser.Expect(TokenType.Register);
        Register = token.Lexeme;
    }

    public override void Compile(Compiler compiler)
    {
        throw new NotImplementedException();
    }

    public override string ToString()
    {
        return $"Register({Register})";
    }

    public override int GetSize()
    {
        throw new NotImplementedException(); // it is invalid to ask for the size of a register node.
    }
}

File: src/compiling/parser/register_node.rs
use crate::compiling::{compiler::Compiler, error_handler::{CompilerError, ErrorCode}, lexer::TokenType, Register};

use super::{Node, Parser};

#[derive(Debug)]
pub struct RegisterNode(pub Register);

impl Node for RegisterNode {
    fn populate(parser: &mut Parser) -> Result<RegisterNode, CompilerError> {
        match &parser.advance().token_type {
            TokenType::Register(reg) => Ok(RegisterNode(reg.clone())),
            _ => {
                Err(CompilerError::from_token(
                    ErrorCode::ExpectedButFound("Register".to_string(), parser.current().token_type.clone()), 
                    parser.current(),
                    false
                ))
            }
        }
    }

    fn get_size(&self) -> i32 {
        panic!("Requesting the size of a register is not a valid operation")
    }
    
    fn compile(&self, _compiler: &mut Compiler) {
        panic!("Compiling a RegisterNode is not a valid operation")
    }
}